harness/determined/common/api/authentication.py

```diff
@@ -104,7 +104,9 @@ def _init_session(
             assert session_user
             token = util.get_det_user_token_from_env()
 
-        if token is not None:
+        # Do not return the stored token if a password is provided.
+        # Authenticate in do_login().
+        if token is not None and password is None:
```

Comment:
Ok, I spent the last hour relearning this part of our codebase.  Gosh I hate this part of our codebase.

I actually am not sure I agree with changing the existing behavior.

If I'm a user, I feel like if I ran this script over and over:

```python
from determined.experimental import client
client.login(user="determined", password="correct_password")
client.do_something()
```

I would expect it to only actually login once, and use the cached login until the cached login expired.

This _is_ the current behavior.  The current behavior has the (unavoidable?) side-effect that if you change the `password` in the script, the failure isn't noticed until the login session expires.

That seems fine to me.

Reply:
Hm, interesting; I disagree. I would expect `login` to log me in regardless. If the function were named something like `ensure_logged_in` or something like that, I'd expect it to do the more complex behavior it currently does.

Reply:
I can see where you're coming from, but I'm not totally convinced.

If we were going to revisit the current design, now is the time to do it, and we should rope in @azhou-determined who's been working on our SDK API plans.

Reply:
my instinct is to also expect `login` to follow a "create a new session if not exists, otherwise no-op" pattern. if we could somehow detect that the password has changed and then treat that as a new login request, that would be ideal. but i'd expect multiple calls to `login` with the same username/password to be treated as "duplicate" requests in the scope of the session.

i don't have a strong justification in either camp, "login" is a bit up to interpretation. but seeing as how the rest of the python SDK takes a policy of caching and attempting to reduce round trips to the master when possible, i'm inclined to think we should keep the behavior as is.